doc/website/blog/2022-03-31-ent-hooks-examples.md

```diff
@@ -0,0 +1,411 @@
+
+### Offloading Long Computations:
+
+In some cases, we have fields in our database that require long computation, we would like to make sure that this data is always updated once
+a change in the data requires it.
+
+A test for it will look like this:
+```go title="ent/schema/dog_test.go"
+func TestCacheHook(t *testing.T) {
+	ctx := context.Background()
+	cs := cache.NewSyncer()
+	c := enttest.Open(t, dialect.SQLite,
+		"file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1",
+		enttest.WithOptions(ent.CacheSyncer(cs)),
+	)
+	cs.Start(ctx, c)
+	cl := c.Cache.Create().SetWalks(-1).SaveX(ctx)
+	d := c.Dog.Create().SetName("Karashindo").SaveX(ctx)
+	u := c.User.Create().SetName("Yoni").
+		SetCache(cl).
+		AddPets(d).
+		SetConnectionString("mysql://root:pass@localhost:3306)").
+		SaveX(ctx)
+	c.Dog.UpdateOne(d).SetName("Fortuna").ExecX(ctx)
+	cs.Close()
+	cl = u.QueryCache().OnlyX(ctx)
+	require.True(t, cl.Walks > 0)
+}
+```
+This test is a bit more complicated, so let's review it line by line:
+First, I create a new `cache syncer` (remember we have a cache entity in our schema).
+
+This `Cache Syncer` can do heavy computations while running in a different context.
+We create an Ent client and inject our `Cache Syncer`, it is disabled by default.
+After that, we start it and provide it with the Ent client, the reason for that is to allow it to change our data (modify our cache entity).
+We create a few entities and update the dog's name, that will trigger a cache sync.
+Finally, we close the `Cache Syncer`, this will block until all pending hooks are completed.
+The assertion validates a value of "walks" larger than 0 (not -1).
+
+#### Recipe:
+
+Let's add a cache layer for our calculations:
+```go title="ent/schema/cache.go"
+// Cache holds the schema definition for the Cache entity.
+type Cache struct {
+	ent.Schema
+}
+
+// Fields of the Cache.
+func (Cache) Fields() []ent.Field {
+	return []ent.Field{
+		field.Int("walks"),
+	}
+}
+```
+
+Now the hook:
+```go title="ent/schema/dog.go"
+// Hooks of the Dog.
+func (Dog) Hooks() []ent.Hook {
+	return []ent.Hook{
+		hook.If(syncCache,
+			hook.HasOp(ent.OpUpdateOne),
+		),		
+	}
+}
+```
+
+The Sync cache will trigger once a dog is updated.
+
+For injecting the dependency in ent we follow [injecting external dependencies](https://entgo.io/docs/code-gen#external-dependencies)
+section and update our  `entc.go`:
+
+```go title="ent/entc.go"
+func main() {
+	opts := []entc.Option{
+		entc.Dependency(
+			entc.DependencyName("CacheSyncer"),
+			entc.DependencyTypeInfo(&field.TypeInfo{
+				Ident:   "hook.Syncer",
+				PkgPath: "github.com/yonidavidson/ent-hooks-examples/hook",
+			}),
+		),
+	}
+	if err := entc.Generate("./schema", &gen.Config{}, opts...); err != nil {
+		log.Fatal("running ent codegen:", err)
+	}
+}
+```
+
+The interface just exposes the sync command:
+```go title="hook/syncer.go"
+package hook
+
+import "context"
+
+type Syncer interface {
```

Comment:
GoDoc. Our code snippets should show clean Go code as well as our code does :-)

Reply:
Done